### ISSUE TITLE
Old Clangd Installs Accumulating

### ISSUE THREAD

**Issue by cpsauer** [clangd/vscode-clangd#285]
Hello, wonderful clangd folks.

A friend of mine was using the VSCode extension and happened to notice that old versions of clangd were accumulating in `~/Library/Application Support/Code/User/globalStorage/llvm-vs-code-extensions.vscode-clangd/install/`. These were presumably all the old versions installed by the (fantastic) auto-update/install server feature, never cleaned up after they were replaced.

Accumulating old versions is hardly a critical issue or anything--just a half a gb or so of disk. But I figured I should give a friendly heads in case it was unintentional. (We'd wrongly assumed that older versions were getting replaced.)

Thanks for everything!
Chris

P.S. This was on macOS, if the path above wasn't already a dead giveaway :)
P.P.S. I'm going to tag @sam-mccall for discussion since I think he wrote the original version in #5. Hopefully that's okay.

**Comment by sam-mccall** [comment]
Hmm, it's not *terrible*: this is stable versions only, and we don't release too often. OTOH we should avoid growing without bound somehow.

Any solution is going to cost some complexity and we don't want to turn into a package manager, or add annoying prompts and then have to add further settings/prompts to turn them off.

A partial solution that should prevent like a decent compromise:
 - after installing, we'll offer _once_ to delete the previous version (after the user can see the new thing works)
 - we won't scan around for other versions, just the immediate previous one
 - this will only trigger for new installations, so it won't clean up existing mess

I've sent a PR for the node-clangd part, the vscode-clangd/coc-clangd parts are then pretty straightforward

**Comment by cpsauer** [comment]
Hah! Certainly not terrible as is; clangd is pretty excellent overall, I'd say :) He'd just stumbled upon it, and I figured I should note it down for you. (Please do lmk if not helpful--I can have a higher issue threshold in the future.)

That solution sounds reasonable! It's not clear to me that some other quite simple options wouldn't better match user needs, given my mental model, but I certainly don't know the constraints and goals as well as you do. My assumption (possibly wrong) is that most users want to have the clangd extension work out of the box as much as possible, and use the latest server version, with confidence that it's configuring their system well. Some other options that seem like they could match those goals well:
1. Only have one version downloaded via this happy path, replacing old ones when downloaded. If the user is (presumably rarely) going back to test an old version, just let them download and install it; they'll have to know the path for the settings anyway, and it's probably unlikely they'd know that old versions were still in this secret location. Main benefit would be automatically handling the most prevalent case well--and simply.
2. Have an additional "delete all old versions" option in the prompt proposed above, preventing long run accumulation if prompts are missed. This, again, based on the guess that most users want the latest and only the latest, but might miss prompts and not want things to just work without thinking about this folder. Seems worse than (1) on the "just works" score, but maybe better if we anticipate lots of people wanting to simultaneously use older version.

^ Obv your call. Just thought I'd write out the options that went through my head.

**Comment by sam-mccall** [comment]
No, thanks for filing - we did consider closing it without a fix but decided it's worthwhile if not too complex.

I don't want to offer multiple options, that's too much support surface for too little benefit I think.
Deleting all the directories is a reasonable alternative, it's harder to communicate to the user though and a bit harder for them to verify it's safe. The main advantage is cleaning up the 3-4 versions users could have installed since we added auto-install.
We can avoid missed prompts: the prompt goes away forever once the user chooses yes/no, but if they dismiss/ignore it it'll prompt again next time.
